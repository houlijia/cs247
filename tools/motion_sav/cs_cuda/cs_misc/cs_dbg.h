#ifndef __CS_DBG_H__ 
#define __CS_DBG_H__ 

#include "cs_header.h"

// #define CUDA_DBG 

int dbg_init( int ) ;
void dbg_clear_buf( int *cp, int size ) ;
void dbg_set_buf( int *cp, int size, int set ) ;
int dbg_put_d_data ( char *dp, char *hp, int size )  ;
int dbg_get_d_data ( char *dp, char *hp, int size )  ;
int dbg_copy_d_data ( char *dtp, char *dfp, int size ) ;
void dbg_pdata_i( const char *s, const int *dp, int size ) ;
void dbg_pdata_f( const char *s, const float *dp, int size ) ;
void dbg_pdata_c( const char *s, const char *dp, int size ) ;
void dbg_pdata_d( const char *s, const float *dp, int size ) ;
void dbg_mdata( int *dp, int size ) ;
void dbg_p_d_data_c ( const char *s, char *dp, int size ) ;
void dbg_p_d_data_i ( const char *s, int *dp, int size ) ;
void dbg_p_d_data_f ( const char *s, float *dp, int size ) ;
void dbg_p_d_data_d ( const char *s, float *dp, int size ) ;
char *dbg_d_malloc_c ( int size ) ;
int *dbg_d_malloc_i ( int size ) ;
float *dbg_d_malloc_f ( int size ) ;

void dbg_p_d_data_f_mn_skip ( const char *s, float *dp, int size, int m, int n, int z,
	int doprint, int perm_size ) ;
void dbg_p_d_data_i_mn_skip ( const char *s, int *dp, int size, int m, int n, int z,
	int doprint, int perm_size ) ;

void dbg_p_d_data_f_mn ( const char *s, float *dp, int size, int m, int n, int doprint ) ;
void dbg_p_d_data_f_mn ( const char *s, float *dp, int size, int m, int n, int doprint, int printrow ) ;
void dbg_p_d_data_i_mn ( const char *s, int *dp, int size, int m, int n, int doprint ) ;
void dbg_p_d_data_c_mn ( const char *s, char *dp, int size, int m, int n, int doprint ) ;
void dbg_p_data_i_mn ( const char *s, int *dp, int size, int m, int n, int doprint ) ;

// print the motion detection T/V/H/Va format
void dbg_p_data_md_f_mn ( const char *s, int *dp, int size, int m, int n, int doprint ) ;

void dbg_p_d_data_f_mn_v2 ( const char *s, float *devp, int size, int doprint,
	struct cube *dp, int blk_in_x, int blk_in_y ) ;
void dbg_p_d_data_i_mn_v2 ( const char *s, int *devp, int size, int doprint,
	struct cube *dp, int blk_in_x, int blk_in_y ) ;
void dbg_p_data_i_mn_v2 ( const char *s, int *hp, int size, int doprint,
	struct cube *dp, int blk_in_x, int blk_in_y ) ;

void dbg_p_d_data_i_cube ( const char *s, int *devp, int vx, int hy, int tz ) ;
void dbg_p_d_data_i_cube ( const char *s, float *devp, int vx, int hy, int tz ) ;
void dbg_p_data_i_cube ( const char *s, float *dp, int vx, int hy, int tz ) ;

void dbg_pr_first_last ( char *s, float *d_p, int len, int pr_size ) ;
void dbg_pr_first_last ( char *s, int *d_p, int len, int pr_size ) ;
void dbg_pr_h_first_last ( char *s, int *d_p, int len, int pr_size ) ;
void dbg_pr_h_first_last ( char *s, char *h_p, int len, int pr_size ) ;

// tsting ...
void dbg_p_d_data_ll ( const char *s, long long *dp, int size ) ;

// for random
int dbg_ck_unique ( char *s, int *dp, int size ) ;
int dbg_perm_ck ( int callid, int callid2, int zero_in_zero, int *d_bp, int size ) ;

// TVH
void cs_p_d_tvh ( const char *s, int *dmem, int record_size, int num_rec, int do_print, int do_skip ) ;
void cs_p_d_tvh ( const char *s, int *dmem, int record_size, int num_rec, int do_print,
	int do_skip, int start ) ;

#endif 
