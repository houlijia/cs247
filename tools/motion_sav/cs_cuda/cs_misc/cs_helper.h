#ifndef __CS_HELPER_H__
#define __CS_HELPER_H__

int max_log2( int ) ;

// clear the device mem ... n in size of int
void set_device_mem_i ( int *d_a, int n, int val ) ;
void clear_device_mem_i ( int *d_a, int n ) ;

void set_device_mem ( float *d_a, int n, float val ) ;
void clear_device_mem ( float *d_a, int n )	 ;

// clear the device mem ... n in size of char
void clear_device_mem_c ( char *d_a, int n ) ;
void set_device_mem_c ( char *d_a, int n, char c ) ;

void h_expand_c_to_i ( char *din, int *dout, int n ) ;	
void h_expand_c_to_i ( unsigned char *din, int *dout, int n ) ;	
template<typename F, typename T> void h_cast_T_to_T ( F *din, T *dout, int n ) ;

// block adjustment
void h_block_adj( int n, int nThreadsPerBlock, int *block );
//

// omp timer 
int omp_timer_init( int ) ;
int omp_timer_on( int ) ;
int omp_timer_off( int ) ;
int omp_timer_get( int, double *dp, int *cp, double *ddp ) ;

// opt
int get_nums( int ac, char *av[], int idx, int cnt, int *np ) ;
unsigned int htoi(const char s[]) ;

// cs timer 
int cs_timer_init( int cnt ) ;
int cs_timer_on( int idx ) ;
int cs_timer_off( int idx ) ;
int cs_timer_get( int idx, clock_t *sp, clock_t *up, int *cp,
	double *asp, double *aup ) ;

// swap
void htonl_device_mem_i ( int *d_a, int n )	;

// tsting ...
void h_tst_longlong ( long long *d_a, int n )	 ;

// abs a vector

int h_set_abs ( int *d_a, int n, int record_len, int skip )	 ;

// set config
void h_set_config( struct cs_xyz *d_a, struct cube *hp ) ;
int h_set_cube_config( struct cube *d_a, struct cube *hp ) ;
void h_p_d_cube_config ( const char *, struct cube *dp ) ;
void h_p_h_cube_config ( const char *, struct cube *dp ) ;

// weight shift
int weight_sft( int scheme, int size, int m, int n ) ;

// generic read/write int from/to device ... for ret code
int put_d_data_i ( int *dp, int *hp, int size ) ;
int put_d_data_f ( float *dp, float *hp, int size ) ;	// size in bytes
int get_d_data_i ( int *dp, int *hp, int size ) ;
int get_d_data_f ( float *dp, float *hp, int size ) ;	// size in bytes

void h_do_int_to_float ( int *intp, float *fp, int tbl_size ) ;

#endif 
