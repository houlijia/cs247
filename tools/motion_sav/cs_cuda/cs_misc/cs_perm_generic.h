#ifndef __CS_PERM_GENERIC_H__
#define __CS_PERM_GENERIC_H__

void
h_do_permutation_generic_f1 ( int *d_input, int *d_output, int *d_perm_tbl,
	int tbl_size ) ;

void
h_do_permutation_generic_f2 ( int *d_input, int *d_output, int *d_perm_tbl,
	int tbl_size ) ;

void
h_do_permutation_generic_f1 ( float *d_input, float *d_output, int *d_perm_tbl,
	int tbl_size ) ;

void
h_do_permutation_generic_f2 ( float *d_input, float *d_output, int *d_perm_tbl,
	int tbl_size ) ;

void
h_do_permutation_generic_inverse ( int *d_output, int *d_perm_tbl,
	int tbl_size ) ;

#endif 
