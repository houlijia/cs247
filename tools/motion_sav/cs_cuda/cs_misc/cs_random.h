#ifndef __CS_RANDOM_H__
#define __CS_RANDOM_H__

int h_set_random_table ( RndC_uint32 seed, RndC_uint32 *h_in, RndC_uint32 *d_out, int size,
	int slot_one_reserved, int do_init )	 ;

#endif 
