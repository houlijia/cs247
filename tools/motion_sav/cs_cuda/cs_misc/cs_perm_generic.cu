#include <sys/types.h>
#include <sys/stat.h>
#include <fcntl.h>
#include <stdio.h>
#include <errno.h>
#include <unistd.h>
#include <stdlib.h>
#include <arpa/inet.h>
#include <math.h>
#include "cs_dbg.h"
#include "cs_cuda.h"
#include "cs_helper.h"

#include "cs_perm_generic.h"

// #define CUDA_DBG
// #define CUDA_DBG1

__global__ void d_do_permutation_generic_f1 ( int *input, int *output,
	int *idxp, int tbl_size )
{
	int t_idx = blockIdx.x*blockDim.x + threadIdx.x;

	while ( t_idx < tbl_size )
	{
		output[ idxp[ t_idx ]] = input[ t_idx ] ;  

		t_idx += CUDA_MAX_THREADS ;
	}
}

/*
	perform

	target(per(i))=orig(i)

	target: outcome vector
	orig: the original vector
	per: permutation vector
*/
void
h_do_permutation_generic_f1 ( int *d_input, int *d_output, int *d_perm_tbl,
	int tbl_size )
{
	int nThreadsPerBlock = CUDA_MAX_THREADS_P_BLK ;
	int nBlocks ; // = ( tbl_size + ( nThreadsPerBlock - 1 ))/nThreadsPerBlock ;

#ifdef CUDA_DBG1 
	printf("%s: din %p dout %p perm %p tblsize %d\n", __func__,
		d_input, d_output, d_perm_tbl, tbl_size ) ;
#endif 

	h_block_adj ( tbl_size, nThreadsPerBlock, &nBlocks ) ;

	d_do_permutation_generic_f1 <<< nBlocks, nThreadsPerBlock >>> ( d_input,
		d_output, d_perm_tbl, tbl_size ) ; 

	cudaThreadSynchronize() ;

#ifdef CUDA_OBS 
	dbg_p_d_data_i("h_do_permutation_generic_f1 perm before", d_input, tbl_size ) ; 
	dbg_p_d_data_i("h_do_permutation_generic_f1 perm after", d_output, tbl_size ) ; 
#endif 
}

__global__ void d_do_permutation_generic_f2 ( int *input, int *output,
	int *idxp, int tbl_size )
{
	int t_idx = blockIdx.x*blockDim.x + threadIdx.x;

	while ( t_idx < tbl_size )
	{
		output[t_idx] = input[ idxp[t_idx]] ;  

		t_idx += CUDA_MAX_THREADS ;
	}
}

/*
	perform

	target(i)=orig((per(i))

	target: outcome vector
	orig: the original vector
	per: permutation vector
*/
void
h_do_permutation_generic_f2 ( int *d_input, int *d_output, int *d_perm_tbl,
	int tbl_size )
{
	int nThreadsPerBlock = CUDA_MAX_THREADS_P_BLK ;
	int nBlocks ; // = ( tbl_size + ( nThreadsPerBlock - 1 ))/nThreadsPerBlock ;

#ifdef CUDA_DBG1 
	printf("%s: din %p dout %p perm %p size %d \n", __func__,
		d_input, d_output, d_perm_tbl, tbl_size ) ;
#endif 

	h_block_adj ( tbl_size, nThreadsPerBlock, &nBlocks ) ;

	d_do_permutation_generic_f2 <<< nBlocks, nThreadsPerBlock >>> ( d_input,
		d_output, d_perm_tbl, tbl_size ) ; 

	cudaThreadSynchronize() ;

#ifdef CUDA_OBS 
	dbg_p_d_data_i("h_do_permutation_generic_f2 perm before", d_input, tbl_size ) ; 
	dbg_p_d_data_i("h_do_permutation_generic_f2 perm after", d_output, tbl_size ) ; 
#endif 
}

// same logic as above but the data types are float

__global__ void d_do_permutation_generic_f1 ( float *input, float *output,
	int *idxp, int tbl_size )
{
	int t_idx = blockIdx.x*blockDim.x + threadIdx.x;

	while ( t_idx < tbl_size )
	{
		output[ idxp[ t_idx ]] = input[ t_idx ] ;  

		t_idx += CUDA_MAX_THREADS ;
	}
}

/*
	perform

	target(per(i))=orig(i)

	target: outcome vector
	orig: the original vector
	per: permutation vector
*/
void
h_do_permutation_generic_f1 ( float *d_input, float *d_output, int *d_perm_tbl,
	int tbl_size )
{
	int nThreadsPerBlock = CUDA_MAX_THREADS_P_BLK ;
	int nBlocks ; // = ( tbl_size + ( nThreadsPerBlock - 1 ))/nThreadsPerBlock ;

#ifdef CUDA_DBG1 
	printf("%s: din %p dout %p perm %p tblsize %d\n", __func__,
		d_input, d_output, d_perm_tbl, tbl_size ) ;
#endif 

	h_block_adj ( tbl_size, nThreadsPerBlock, &nBlocks ) ;

	d_do_permutation_generic_f1 <<< nBlocks, nThreadsPerBlock >>> ( d_input,
		d_output, d_perm_tbl, tbl_size ) ; 

	cudaThreadSynchronize() ;

#ifdef CUDA_OBS 
	dbg_p_d_data_f("h_do_permutation_generic_f1 perm before", d_input, tbl_size ) ; 
	dbg_p_d_data_f("h_do_permutation_generic_f1 perm after", d_output, tbl_size ) ; 
#endif 
}

__global__ void d_do_permutation_generic_f2 ( float *input, float *output,
	int *idxp, int tbl_size )
{
	int t_idx = blockIdx.x*blockDim.x + threadIdx.x;

	while ( t_idx < tbl_size )
	{
		output[t_idx] = input[ idxp[t_idx]] ;  

		t_idx += CUDA_MAX_THREADS ;
	}
}

/*
	perform

	target(i)=orig((per(i))

	target: outcome vector
	orig: the original vector
	per: permutation vector
*/
void
h_do_permutation_generic_f2 ( float *d_input, float *d_output, int *d_perm_tbl,
	int tbl_size )
{
	int nThreadsPerBlock = CUDA_MAX_THREADS_P_BLK ;
	int nBlocks ; // = ( tbl_size + ( nThreadsPerBlock - 1 ))/nThreadsPerBlock ;

#ifdef CUDA_DBG1 
	printf("%s: din %p dout %p perm %p size %d tblsize %d\n", __func__,
		d_input, d_output, d_perm_tbl, tbl_size ) ;
#endif 

	h_block_adj ( tbl_size, nThreadsPerBlock, &nBlocks ) ;

	d_do_permutation_generic_f2 <<< nBlocks, nThreadsPerBlock >>> ( d_input,
		d_output, d_perm_tbl, tbl_size ) ; 

	cudaThreadSynchronize() ;

#ifdef CUDA_OBS 
	dbg_p_d_data_f("h_do_permutation_generic_f2 perm before", d_input, tbl_size ) ; 
	dbg_p_d_data_f("h_do_permutation_generic_f2 perm after", d_output, tbl_size ) ; 
#endif 
}

__global__ void d_do_permutation_generic_inverse ( int *output,
	int *idxp, int tbl_size )
{
	int t_idx = blockIdx.x*blockDim.x + threadIdx.x;

	while ( t_idx < tbl_size )
	{
		output[ idxp[ t_idx ]] = t_idx ;  

		t_idx += CUDA_MAX_THREADS ;
	}
}

/*
	perform

	target(per(i))=(i)

	target: outcome vector
	orig: the original vector
	per: permutation vector
*/
void
h_do_permutation_generic_inverse ( int *d_output, int *d_perm_tbl,
	int tbl_size )
{
	int nThreadsPerBlock = CUDA_MAX_THREADS_P_BLK ;
	int nBlocks ; // = ( tbl_size + ( nThreadsPerBlock - 1 ))/nThreadsPerBlock ;

#ifdef CUDA_DBG1 
	printf("%s: dout %p perm %p tblsize %d\n", __func__,
		d_output, d_perm_tbl, tbl_size ) ;
#endif 

	h_block_adj ( tbl_size, nThreadsPerBlock, &nBlocks ) ;

	d_do_permutation_generic_inverse <<< nBlocks, nThreadsPerBlock >>> (
		d_output, d_perm_tbl, tbl_size ) ; 

	cudaThreadSynchronize() ;

#ifdef CUDA_OBS 
	dbg_p_d_data_i("h_do_permutation_generic_inverse perm before", d_perm_tbl, tbl_size ) ; 
	dbg_p_d_data_i("h_do_permutation_generic_inverse perm after", d_output, tbl_size ) ; 
#endif 
}
