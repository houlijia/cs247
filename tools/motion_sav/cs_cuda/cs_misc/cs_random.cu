#include <sys/types.h>
#include <sys/stat.h>
#include <fcntl.h>
#include <stdio.h>
#include <errno.h>
#include <unistd.h>
#include <stdlib.h>
#include <arpa/inet.h>
#include <math.h>

#include "cs_dbg.h"
#include "cs_cuda.h"
#include "cs_helper.h"

#include "RndCState.h"
#include "RndC_ifc.h"

#include "cs_random.h"

// #define CUDA_DBG
#define CUDA_DBG1

static RndCState cs_random_state ;

// LDL stop here ... in SensingMatrixWH.makepuermutations
// PL = 1 .. 1 + random(order - 1) 
// PR = random(order) 


// val can be either positive or negative

__global__ void 
d_set_random_table( RndC_uint32 *fp, int size, RndC_uint32 val )
{
	int tid = blockIdx.x*blockDim.x + threadIdx.x;

	while ( tid < size )
	{
		fp[ tid ] += val ;
		tid += CUDA_MAX_THREADS ;
	} 
}

/*
    seed : the seed for random number generation
	h_in: points to the host buffer before download to device, if null, a temp buffer will be
		allocated and deleted when done.
	slot_one_reserved: if set [0] will have 0
*/

int
h_set_random_table ( RndC_uint32 seed, RndC_uint32 *h_in, RndC_uint32 *d_out, int size,
	int slot_one_reserved, int do_init )	
{
	int nThreadsPerBlock = CUDA_MAX_THREADS_P_BLK ;
	int nBlocks ;
	RndC_uint32 *h_in_memp = NULL ;
	int i, need_adjustment = 0 ;

	if ( h_in == NULL )
	{
		h_in_memp = ( RndC_uint32 *)malloc ( sizeof ( RndC_uint32 ) * size ) ;
		if ( h_in_memp == NULL )
			return ( 0 ) ;
	} else
		h_in_memp = h_in ;


	if ( do_init )
		init_RndC( &cs_random_state, seed ) ;

	if ( slot_one_reserved )
	{
		h_in_memp[0] = ( RndC_uint32 )0 ;
		randperm1_RndC( &cs_random_state, size - 1, h_in_memp+1 ) ; // from 1 to ( size - 1 )
	} else
	{
		randperm1_RndC( &cs_random_state, size, h_in_memp ) ; // from 1 to ( size )
			// need to subtract one 
		need_adjustment++ ;
	}

	if (( i = cudaMemcpy( d_out, h_in_memp, sizeof ( RndC_uint32 ) * size,
		cudaMemcpyHostToDevice)) != cudaSuccess )
	{
		printf("%s: download cp fail: %d\n", __func__, i ) ;
		
		if ( h_in == NULL )
			free ( h_in_memp ) ;
		return ( 0 ) ;
	}

	if ( need_adjustment )
	{
		h_block_adj ( size, nThreadsPerBlock, &nBlocks ) ;

		d_set_random_table <<< nBlocks, nThreadsPerBlock >>> ( d_out, size, ( RndC_uint32) -1 ) ;

		cudaThreadSynchronize() ;
	}

	if ( h_in == NULL )
		free ( h_in_memp ) ;

	return ( 1 ) ;
}

