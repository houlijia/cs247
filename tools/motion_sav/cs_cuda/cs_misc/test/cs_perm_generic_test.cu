#include <stdio.h>

#include "cs_dbg.h"
#include "cs_helper.h"
#include "cs_perm_generic.h"

#include "RndCState.h"
#include "RndC_ifc.h"

#include "cs_random.h"

#define TEST_SIZE 8
#define BUF_LOG_2  10

int orig[] = { 11, 22, 33, 44, 55, 66, 77, 88 } ;
float orig1[] = { 11.0, 22.0, 33.0, 44.0, 55.0, 66.0, 77.0, 88.0 } ;
int perm[] = { 1, 4, 3, 5, 7, 0, 2, 6 } ;

int tst_2() ;

main()
{
	int k ;
	int *tp, *pp, *op ;
	float *ftp, *fpp, *fop ;

	dbg_init( 102400 ) ;

	if (( k = cudaMalloc( &pp, sizeof ( int ) * TEST_SIZE * 3 )) != cudaSuccess )
	{
		printf("%s: alloc failed %d \n", __func__, k ) ;
		exit ( 0 ) ;
	}

	op = pp + TEST_SIZE ;
	tp = op + TEST_SIZE ;

	put_d_data_i ( op, orig, sizeof(int) * TEST_SIZE ) ;
	put_d_data_i ( pp, perm, sizeof(int) * TEST_SIZE ) ;

	h_do_permutation_generic_f1 ( op, tp, pp, TEST_SIZE ) ;

	dbg_p_d_data_i ("f1", tp, TEST_SIZE ) ; 

	/* should see
	   0 -- 00000042 66
	   1 -- 0000000b 11
	   2 -- 0000004d 77
	   3 -- 00000021 33
	   4 -- 00000016 22
	   5 -- 0000002c 44
	   6 -- 00000058 88
	   7 -- 00000037 55
	*/

	h_do_permutation_generic_f2 ( op, tp, pp, TEST_SIZE ) ;

	dbg_p_d_data_i ("f2", tp, TEST_SIZE ) ; 

	/* should see
	   0 -- 00000016 22
	   1 -- 00000037 55
	   2 -- 0000002c 44
	   3 -- 00000042 66
	   4 -- 00000058 88
	   5 -- 0000000b 11
	   6 -- 00000021 33
	   7 -- 0000004d 77
	 */

	// float
	if (( k = cudaMalloc( &fpp, sizeof ( float ) * TEST_SIZE * 3 )) != cudaSuccess )
	{
		printf("%s: alloc failed %d \n", __func__, k ) ;
		exit ( 0 ) ;
	}

	fop = fpp + TEST_SIZE ;
	ftp = fop + TEST_SIZE ;

	put_d_data_f ( fop, orig1, sizeof(float) * TEST_SIZE ) ;
	put_d_data_i ( pp, perm, sizeof(int) * TEST_SIZE ) ;

	h_do_permutation_generic_f1 ( fop, ftp, pp, TEST_SIZE ) ;

	dbg_p_d_data_f ("float f1", ftp, TEST_SIZE ) ; 

	/* should see
	   0 -- 00000042 66
	   1 -- 0000000b 11
	   2 -- 0000004d 77
	   3 -- 00000021 33
	   4 -- 00000016 22
	   5 -- 0000002c 44
	   6 -- 00000058 88
	   7 -- 00000037 55
	*/

	h_do_permutation_generic_f2 ( fop, ftp, pp, TEST_SIZE ) ;

	dbg_p_d_data_f ("float f2", ftp, TEST_SIZE ) ; 

	/* should see
	   0 -- 00000016 22
	   1 -- 00000037 55
	   2 -- 0000002c 44
	   3 -- 00000042 66
	   4 -- 00000058 88
	   5 -- 0000000b 11
	   6 -- 00000021 33
	   7 -- 0000004d 77
	 */

	tst_2() ;

}

int
tst_2()
{
	// set up random numbers buffer

	int i, *t, *ot, *d_pL, *d_pR, *d_t1, *d_t2, *d_t3, *d_t4, *d_t5 ;

	printf("starting tst_2 ... \n") ;

	if (( i = cudaMalloc ( &d_pL, BUF_LOG_2 * sizeof( int ) * 7  )) != cudaSuccess )
	{
#ifdef CUDA_DBG 
		printf("%s: malloc failed %d \n", __func__, i ) ;
#endif 
		return ( 0 ) ;
	}

	d_pR = d_pL + BUF_LOG_2 ;
	d_t1 = d_pR + BUF_LOG_2 ;
	d_t2 = d_t1 + BUF_LOG_2 ;
	d_t3 = d_t2 + BUF_LOG_2 ;
	d_t4 = d_t3 + BUF_LOG_2 ;
	d_t5 = d_t4 + BUF_LOG_2 ;

	// set up orig data to be 0..9

	t = (int *)malloc ( BUF_LOG_2 * sizeof ( int )) ;

	ot = t ;

	for ( i = 0 ; i < BUF_LOG_2 ; i++ )
		*t++ = i ;

	put_d_data_i ( d_t1, ot, sizeof(int) * BUF_LOG_2 ) ;

	// set up random table

	h_set_random_table( 0, NULL, ( RndC_uint32 * )d_pL, BUF_LOG_2, 1, 1 ) ;
	h_set_random_table( 1, NULL, ( RndC_uint32 * )d_pR, BUF_LOG_2, 0, 1 ) ;

	dbg_p_d_data_i ( "d_pL", ( int *)d_pL, BUF_LOG_2 ) ;
	dbg_p_d_data_i ( "d_pR", ( int *)d_pR, BUF_LOG_2 ) ;

	// ok 

	dbg_p_d_data_i ( "d_t1", ( int *)d_t1, BUF_LOG_2 ) ;
	h_do_permutation_generic_f1( d_t1, d_t2, d_pR, BUF_LOG_2 ) ; 
	dbg_p_d_data_i ( "d_t2", ( int *)d_t2, BUF_LOG_2 ) ;

	h_do_permutation_generic_f2( d_t2, d_t3, d_pL, BUF_LOG_2 ) ; 
	dbg_p_d_data_i ( "d_t3", ( int *)d_t3, BUF_LOG_2 ) ;

	h_do_permutation_generic_f1( d_t3, d_t4, d_pL, BUF_LOG_2 ) ; 
	dbg_p_d_data_i ( "d_t4", ( int *)d_t4, BUF_LOG_2 ) ;

	h_do_permutation_generic_f2( d_t4, d_t2, d_pR, BUF_LOG_2 ) ; 
	dbg_p_d_data_i ( "d_t2", ( int *)d_t2, BUF_LOG_2 ) ;

	h_do_permutation_generic_inverse ( d_t5, d_pL, BUF_LOG_2 ) ;
	dbg_p_d_data_i ( "d_t5", ( int *)d_t5, BUF_LOG_2 ) ;

	return ( 1 ) ;
}
