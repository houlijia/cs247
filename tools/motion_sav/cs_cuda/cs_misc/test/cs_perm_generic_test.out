dbg_p_d_data_i: f1 size 32 dp 0x1300200040
dbg_pdata_i: f1
0 -- 00000042 66
1 -- 0000000b 11
2 -- 0000004d 77
3 -- 00000021 33
4 -- 00000016 22
5 -- 0000002c 44
6 -- 00000058 88
7 -- 00000037 55
dbg_p_d_data_i: f2 size 32 dp 0x1300200040
dbg_pdata_i: f2
0 -- 00000016 22
1 -- 00000037 55
2 -- 0000002c 44
3 -- 00000042 66
4 -- 00000058 88
5 -- 0000000b 11
6 -- 00000021 33
7 -- 0000004d 77
dbg_p_d_data_f: h_do_permutation_generic_f1 perm before size 32 dp 0x1300200220
dbg_pdata_f: h_do_permutation_generic_f1 perm before
0 -- 11.000000
1 -- 22.000000
2 -- 33.000000
3 -- 44.000000
4 -- 55.000000
5 -- 66.000000
6 -- 77.000000
7 -- 88.000000
dbg_p_d_data_f: h_do_permutation_generic_f1 perm after size 32 dp 0x1300200240
dbg_pdata_f: h_do_permutation_generic_f1 perm after
0 -- 66.000000
1 -- 11.000000
2 -- 77.000000
3 -- 33.000000
4 -- 22.000000
5 -- 44.000000
6 -- 88.000000
7 -- 55.000000
dbg_p_d_data_f: float f1 size 32 dp 0x1300200240
dbg_pdata_f: float f1
0 -- 66.000000
1 -- 11.000000
2 -- 77.000000
3 -- 33.000000
4 -- 22.000000
5 -- 44.000000
6 -- 88.000000
7 -- 55.000000
dbg_p_d_data_f: float f2 size 32 dp 0x1300200240
dbg_pdata_f: float f2
0 -- 22.000000
1 -- 55.000000
2 -- 44.000000
3 -- 66.000000
4 -- 88.000000
5 -- 11.000000
6 -- 33.000000
7 -- 77.000000
dbg_p_d_data_i: IN h_set_random_table size 40 dp 0x1300200400
dbg_pdata_i: IN h_set_random_table
0 -- 00000000 0
1 -- 00000006 6
2 -- 00000003 3
3 -- 00000007 7
4 -- 00000008 8
5 -- 00000005 5
6 -- 00000001 1
7 -- 00000002 2
8 -- 00000004 4
9 -- 00000009 9
dbg_p_d_data_i: IN h_set_random_table size 40 dp 0x1300200428
dbg_pdata_i: IN h_set_random_table
0 -- 00000003 3
1 -- 00000006 6
2 -- 00000005 5
3 -- 00000007 7
4 -- 00000004 4
5 -- 00000008 8
6 -- 00000009 9
7 -- 00000001 1
8 -- 0000000a 10
9 -- 00000002 2
dbg_p_d_data_i: IN h_set_random_table size 40 dp 0x1300200428
dbg_pdata_i: IN h_set_random_table
0 -- 00000002 2
1 -- 00000005 5
2 -- 00000004 4
3 -- 00000006 6
4 -- 00000003 3
5 -- 00000007 7
6 -- 00000008 8
7 -- 00000000 0
8 -- 00000009 9
9 -- 00000001 1
dbg_p_d_data_i: d_pL size 40 dp 0x1300200400
dbg_pdata_i: d_pL
0 -- 00000000 0
1 -- 00000006 6
2 -- 00000003 3
3 -- 00000007 7
4 -- 00000008 8
5 -- 00000005 5
6 -- 00000001 1
7 -- 00000002 2
8 -- 00000004 4
9 -- 00000009 9
dbg_p_d_data_i: d_pR size 40 dp 0x1300200428
dbg_pdata_i: d_pR
0 -- 00000002 2
1 -- 00000005 5
2 -- 00000004 4
3 -- 00000006 6
4 -- 00000003 3
5 -- 00000007 7
6 -- 00000008 8
7 -- 00000000 0
8 -- 00000009 9
9 -- 00000001 1
dbg_p_d_data_i: d_t1 size 40 dp 0x1300200450
dbg_pdata_i: d_t1
0 -- 00000000 0
1 -- 00000001 1
2 -- 00000002 2
3 -- 00000003 3
4 -- 00000004 4
5 -- 00000005 5
6 -- 00000006 6
7 -- 00000007 7
8 -- 00000008 8
9 -- 00000009 9
dbg_p_d_data_i: d_t2 size 40 dp 0x1300200478
dbg_pdata_i: d_t2
0 -- 00000007 7
1 -- 00000009 9
2 -- 00000000 0
3 -- 00000004 4
4 -- 00000002 2
5 -- 00000001 1
6 -- 00000003 3
7 -- 00000005 5
8 -- 00000006 6
9 -- 00000008 8
dbg_p_d_data_i: d_t3 size 40 dp 0x13002004a0
dbg_pdata_i: d_t3
0 -- 00000007 7
1 -- 00000003 3
2 -- 00000004 4
3 -- 00000005 5
4 -- 00000006 6
5 -- 00000001 1
6 -- 00000009 9
7 -- 00000000 0
8 -- 00000002 2
9 -- 00000008 8
dbg_p_d_data_i: d_t4 size 40 dp 0x13002004c8
dbg_pdata_i: d_t4
0 -- 00000007 7
1 -- 00000009 9
2 -- 00000000 0
3 -- 00000004 4
4 -- 00000002 2
5 -- 00000001 1
6 -- 00000003 3
7 -- 00000005 5
8 -- 00000006 6
9 -- 00000008 8
dbg_p_d_data_i: d_t2 size 40 dp 0x1300200478
dbg_pdata_i: d_t2
0 -- 00000000 0
1 -- 00000001 1
2 -- 00000002 2
3 -- 00000003 3
4 -- 00000004 4
5 -- 00000005 5
6 -- 00000006 6
7 -- 00000007 7
8 -- 00000008 8
9 -- 00000009 9
dbg_p_d_data_i: d_t5 size 40 dp 0x13002004f0
dbg_pdata_i: d_t5
0 -- 00000000 0
1 -- 00000006 6
2 -- 00000007 7
3 -- 00000002 2
4 -- 00000008 8
5 -- 00000005 5
6 -- 00000001 1
7 -- 00000003 3
8 -- 00000004 4
9 -- 00000009 9
h_do_permutation_generic_f1: din 0x1300200020 dout 0x1300200040 perm 0x1300200000 tblsize 8
h_do_permutation_generic_f2: din 0x1300200020 dout 0x1300200040 perm 0x1300200000 size 8 
h_do_permutation_generic_f1: din 0x1300200220 dout 0x1300200240 perm 0x1300200000 tblsize 8
h_do_permutation_generic_f2: din 0x1300200220 dout 0x1300200240 perm 0x1300200000 size 8 tblsize 23383152
starting tst_2 ... 
h_do_permutation_generic_f1: din 0x1300200450 dout 0x1300200478 perm 0x1300200428 tblsize 10
h_do_permutation_generic_f2: din 0x1300200478 dout 0x13002004a0 perm 0x1300200400 size 10 
h_do_permutation_generic_f1: din 0x13002004a0 dout 0x13002004c8 perm 0x1300200400 tblsize 10
h_do_permutation_generic_f2: din 0x13002004c8 dout 0x1300200478 perm 0x1300200428 size 10 
h_do_permutation_generic_inverse: dout 0x13002004f0 perm 0x1300200400 tblsize 10
